type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(CustomIdent),
+
+	// primitives, including build-in types and common precludes
+	Bool,
+	Str,
+	U8,
+	U16,
+	U32,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	I64,
+	I128,
+	Array(ArrayIdent),
+	Slice(SliceIdent),
+	Tuple(TupleIdent),
+	Option(OptionIdent),
+	Result(ResultIdent),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace(Vec<&'static str>);
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Option<Vec<IdentKind>>,
```

Comment:
`Vec<IdentKind>` should be sufficient. An empty `Vec` represents the `()` type, while all other `Vec` lengths represent their actual tuple. Note that an empty `Vec` doesn't heap allocate so it is actually efficient.

Reply:
I was thinking using `Option` here might make it more distinct between `Tuple` and `Unit`, but it might not be that important anyway, since they're same stuff under the hood, will update.